### PR TITLE
Yacht-1001: SLI for backup Quality

### PR DESCRIPTION
**Pull request by przemyslaw-jasinski**

### CODE REVIEW

File: terraform/SLI_quality_views.tf [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+              SELECT
+                source_project_id, source_dataset_id, source_table_id, source_partition_id,
+                backup_dataset_id, backup_table_id, backup_last_modified, backup_num_bytes, backup_num_rows
+              FROM [${local.SLI_views_destination_project}.${var.SLI_backup_quality_views_dataset}.last_backup_in_census]
+            ) AS last_backup_in_census
+            ON source_table.projectId=last_backup_in_census.source_project_id AND
+               source_table.datasetId=last_backup_in_census.source_dataset_id AND
+               source_table.tableId=last_backup_in_census.source_table_id AND
+               source_table.partitionId=last_backup_in_census.source_partition_id
+            WHERE (source_table.numBytes-last_backup_in_census.backup_num_bytes) != 0
```
Comment (radkomateusz) [inline terraform/SLI_quality_views.tf]: source_table.numBytes  != last_backup_in_census.backup_num_bytes . instead of division

Comment (przemyslaw-jasinski) [reply terraform/SLI_quality_views.tf]: Done

File: terraform/SLI_3_days_views.tf [hunk 2]
```diff
@@ -1,6 +1,6 @@
 resource "google_bigquery_table" "census_data_3_days_ago_view" {
```
Comment (marcin-kolda) [inline terraform/SLI_3_days_views.tf]: We could create a module for those X days SLI :)

File: terraform/vars.tf [hunk 3]
```diff
@@ -1,5 +1,10 @@
 terraform {
   required_version = "> 0.11"
+
+  backend "gcs" {
+    bucket = "dev-atm-data-lake-drain-tf-state"
```
Comment (marcin-kolda) [inline terraform/vars.tf]: Please make it configurable during the initialisation as described in https://www.terraform.io/docs/backends/config.html

File: terraform/SLI_quality_views.tf [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+            ) AS census
+            ON census.datasetId=last_backup.backup_dataset_id AND census.tableId=last_backup.backup_table_id
+        EOF
+    use_legacy_sql = true
+  }
+
+  depends_on = ["google_bigquery_dataset.SLI_backup_quality_views_dataset"]
+}
+
+resource "google_bigquery_table" "SLI_quality" {
```
Comment (marcin-kolda) [inline terraform/SLI_quality_views.tf]: Please add 'description' to all views

File: terraform/SLI_backup_quality_views_dataset.tf [hunk 5]
```diff
@@ -0,0 +1,7 @@
+resource "google_bigquery_dataset" "SLI_backup_quality_views_dataset" {
```
Comment (marcin-kolda) [inline terraform/SLI_backup_quality_views_dataset.tf]: Please add dataset 'description'

File: terraform/SLI_quality_views.tf [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+            #legacySQL
+            SELECT projectId, datasetId, tableId, partitionId, lastModifiedTime, numBytes, numRows FROM (
+              SELECT projectId, datasetId, tableId, 'null' AS partitionId, lastModifiedTime, numBytes, numRows
+              FROM [${var.gcp_census_project}.bigquery_views_legacy_sql.table_metadata_v1_0]
+              WHERE DATEDIFF(CURRENT_TIMESTAMP(), lastModifiedTime) >= 3 AND projectId != "${var.bbq_project}"
+              ), (
+              SELECT projectId, datasetId, tableId, partitionId, lastModifiedTime, numBytes, numRows
+              FROM [${var.gcp_census_project}.bigquery_views_legacy_sql.partition_metadata_v1_0]
+              WHERE DATEDIFF(CURRENT_TIMESTAMP(), lastModifiedTime) >= 3 AND projectId != "${var.bbq_project}"
+            )
```
[the remark this answers is not in the document]
Comment (radkomateusz) [reply terraform/SLI_quality_views.tf]: probably to the first select from table_metadata you ned to add: 

 AND timePartitioning.type IS NULL AND type='TABLE' to get only non partitioned tables from first query

[the remark this answers is not in the document]
Comment (radkomateusz) [reply terraform/SLI_quality_views.tf]: but to use above line you need to query census data directly, not via view

[the remark this answers is not in the document]
Comment (przemyslaw-jasinski) [reply terraform/SLI_quality_views.tf]: Done

File: terraform/SLI_quality_views.tf [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+            #legacySQL
+            SELECT
+              last_backup.source_project_id AS source_project_id,
+              last_backup.source_dataset_id AS source_dataset_id,
+              last_backup.source_table_id AS source_table_id,
+              last_backup.source_partition_id AS source_partition_id,
+              census.datasetId AS backup_dataset_id,
+              census.tableId AS backup_table_id,
+              census.lastModifiedTime as backup_last_modified,
+              census.numBytes AS backup_num_bytes,
```
[the remark this answers is not in the document]
Comment (przemyslaw-jasinski) [reply terraform/SLI_quality_views.tf]: Done

File: terraform/SLI_backup_quality_views_dataset.tf (hunk 5 above)
Comment (przemyslaw-jasinski) [reply terraform/SLI_backup_quality_views_dataset.tf]: Done

File: terraform/SLI_quality_views.tf (hunk 4 above)
Comment (przemyslaw-jasinski) [reply terraform/SLI_quality_views.tf]: Done

File: terraform/vars.tf (hunk 3 above)
Comment (przemyslaw-jasinski) [reply terraform/vars.tf]: Done

File: terraform/SLI_3_days_views.tf (hunk 2 above)
Comment (przemyslaw-jasinski) [reply terraform/SLI_3_days_views.tf]: Decide offline to skip it

**Pull request merged**